File: ruby/198.house-robber.rb
# @param {Integer[]} nums
# @return {Integer}
def rob(nums)
  prev_a = 0
  prev_b = 0
  nums.each do |num|
    a = prev_b + num
    b = [prev_a, prev_b].max

    prev_a = a
    prev_b = b
  end

  [prev_a, prev_b].max
end

File: rust/src/198.house-robber.rs
use crate::Solution;

#[allow(dead_code)]
impl Solution {
    pub fn rob(nums: Vec<i32>) -> i32 {
        let (a, b) = nums.iter().fold((0, 0), |(prev_a, prev_b), &num| {
            (prev_b + num, prev_a.max(prev_b))
        });

        a.max(b)
    }
}
